### ISSUE TITLE
Only detection with val.py produces good results

### ISSUE THREAD

**Issue by shells-above** [ultralytics/yolov5#8291]
### Search before asking

- [x] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Validation

### Bug

Im running into an issue with inferences / validation, whenever I try to do inferences with this: (from the main page)

`model = torch.hub.load('ultralytics/yolov5', 'custom', path='best.pt)
results = model("image.jpg")
results.print()`

I get ~500 ms inference times (fixed) and no detections on my custom model (fixed, but still performs worse compared to val.py), but when I try this again but with one of the default models I get detections.

However, when I try this using val.py with:

`python val.py --weights best.pt --data infested.yaml --img 128 --project infestations --name val_ `

I get ~1.5 ms inference times and the correct detections.

Edit: Just tested detect.py, it is faster than the torch hub way but still doesnt detect anything (fixed). Also, I "fixed" the slow inference times for torch hub. I still dont get any detections though.

### Environment

- yolo: YOLOv5  2022-6-14 torch-1.13.0.dev20220616+cu113 CUDA:0 (NVIDIA TITAN Xp, 12288MiB)
- os: windows 10
- python: Python-3.10.4

**Comment by shells-above** [comment]
Now that I read it a couple more times, this seems more like a question. Sorry about marking it as a bug.

**Comment by glenn-jocher** [comment]
@2xmonth detect.py and hub inference produce identical results with identical arguments, but you're asking here why an apple is not an orange. You've passed completely different argument, i.e. img 128

**Comment by shells-above** [comment]
I fixed the no detection problem on torch hub. But I am still getting worse results. (I re-ran both multiple times)

Prediction with torch hub (size=128):
![tmpi44cmx5y](https://user-images.githubusercontent.com/92888249/175057376-f0c9b71d-d126-4b2e-86ed-698349cfea2e.PNG)

Prediction with val.py: (python val.py --weights best.pt --img 128 --data infested_1.yaml)
![val_batch0_pred](https://user-images.githubusercontent.com/92888249/175057506-36ffbb16-94c2-42ac-bc3e-4657424efea0.jpg)

They are both using the same model, and have the same arguments passed to them. Yet torch hub still performs worse. (Compared to val.py, as you said detect.py performs the same)

**Comment by glenn-jocher** [comment]
@2xmonth 👋 Hello, thanks for asking about the differences between [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py), [detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py) and [val.py](https://github.com/ultralytics/yolov5/blob/master/val.py) in YOLOv5 🚀.

These 3 files are designed for different purposes and utilize different dataloaders with different settings. train.py dataloaders are designed for a speed-accuracy compromise, val.py is designed to obtain the best mAP on a validation dataset, and detect.py is designed for best real-world inference results. A few important aspects of each:


## train.py

- **trainloader** LoadImagesAndLabels(): designed to load train dataset images and labels. Augmentation capable and enabled. https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/train.py#L210-L213 **val_loader** LoadImagesAndLabels(): designed to load val dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/train.py#L220-L223
- image size: `640`
- rectangular inference: `False`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `None`

## val.py

- **dataloader** LoadImagesAndLabels(): designed to load train, val, test dataset images and labels. Augmentation capable but disabled. 
- https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/val.py#L152-L153
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `0.5 * maximum stride`


## detect.py

- **dataloaders** (multiple): designed for loading multiple types of media (images, videos, globs, directories, streams). https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/detect.py#L120-L128
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L152): `0.25` 
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L153): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/utils/general.py#L415): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L45-L54): `None`


## YOLOv5 PyTorch Hub Inference

YOLOv5 PyTorch Hub models are`AutoShape()` instances used for image loading, preprocessing, inference and NMS. For more info see [YOLOv5 PyTorch Hub Tutorial](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)
https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/models/common.py#L276-L282

- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.25 `
- [iou threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L288): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `None`

Good luck 🍀 and let us know if you have any other questions!
